src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+  C = NULL;
+  C = SUNAdaptController_NewEmpty(sunctx);
+  if (C == NULL) { return (NULL); }
+
+  /* Attach operations */
+  C->ops->gettype      = SUNAdaptController_GetType_Soderlind;
+  C->ops->estimatestep = SUNAdaptController_EstimateStep_Soderlind;
+  C->ops->reset        = SUNAdaptController_Reset_Soderlind;
+  C->ops->setdefaults  = SUNAdaptController_SetDefaults_Soderlind;
+  C->ops->write        = SUNAdaptController_Write_Soderlind;
+  C->ops->seterrorbias = SUNAdaptController_SetErrorBias_Soderlind;
+  C->ops->updateh      = SUNAdaptController_UpdateH_Soderlind;
+  C->ops->space        = SUNAdaptController_Space_Soderlind;
+
+  /* Create content */
+  content = NULL;
+  content = (SUNAdaptControllerContent_Soderlind)malloc(sizeof *content);
+  if (content == NULL)
+  {
+    (void) SUNAdaptController_Destroy(C);
+    return (NULL);
+  }
+
+  /* Attach content */
+  C->content = content;
+
+  /* Fill content with default/reset values */
+  SUNAdaptController_SetDefaults_Soderlind(C);
+  SUNAdaptController_Reset_Soderlind(C);
+
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set Soderlind parameters
+ */
+
+int SUNAdaptController_SetParams_Soderlind(SUNAdaptController C,
+                                           sunrealtype k1, sunrealtype k2,
+                                           sunrealtype k3, sunrealtype k4,
+                                           sunrealtype k5)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = k2;
+  SODERLIND_K3(C) = k3;
+  SODERLIND_K4(C) = k4;
+  SODERLIND_K5(C) = k5;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create a PID controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_PID(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PID(C,
+                                          DEFAULT_PID_K1,
+                                          DEFAULT_PID_K2,
+                                          DEFAULT_PID_K3);
+  return (C);
+}
+
+/* -----------------------------------------------------------------
+ * Function to set PID parameters
+ */
+
+int SUNAdaptController_SetParams_PID(SUNAdaptController C,
+                                     sunrealtype k1, sunrealtype k2,
+                                     sunrealtype k3)
+{
+  SODERLIND_K1(C) = k1;
+  SODERLIND_K2(C) = k2;
+  SODERLIND_K3(C) = k3;
+  SODERLIND_K4(C) = RCONST(0.0);
+  SODERLIND_K5(C) = RCONST(0.0);
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+/* -----------------------------------------------------------------
+ * Function to create a PI controller (subset of Soderlind)
+ */
+
+SUNAdaptController SUNAdaptController_PI(SUNContext sunctx)
+{
+  SUNAdaptController C;
+  C = SUNAdaptController_Soderlind(sunctx);
+  (void) SUNAdaptController_SetParams_PI(C,
```

Comment:
It unlikely but we should check this return value, destroy the controller, and return `NULL` if setting the parameters (somehow) fails.

Reply:
See comments below about checking for a `NULL` input